mlflow/pyfunc/__init__.py

```diff
@@ -828,7 +829,48 @@ def _create_model_downloading_tmp_dir(should_use_nfs):
 _MLFLOW_SERVER_OUTPUT_TAIL_LINES_TO_KEEP = 200
 
 
-def spark_udf(spark, model_uri, result_type="double", env_manager=_EnvManager.LOCAL):
+def _convert_spec_to_spark_type(spec):
+    from mlflow.types.schema import ColSpec, TensorSpec, DataType
+    from pyspark.sql.types import ArrayType
+
+    # TODO: handle optional output columns.
+    if isinstance(spec, ColSpec):
+        return spec.type.to_spark()
+    elif isinstance(spec, TensorSpec):
+        data_type = DataType.from_numpy_type(spec.type)
+        if data_type is None:
+            raise ValueError(
+                f"Model output tensor spec type {spec.type} is not supported in spark_udf."
+            )
+
+        if len(spec.shape) == 1:
+            return ArrayType(data_type.to_spark())
+        elif len(spec.shape) == 2:
+            return ArrayType(ArrayType(data_type.to_spark()))
+        else:
+            raise ValueError(
+                "Only one dimension or 2 dimensions tensors are supported as spark_udf "
+                f"return value, but model output '{spec.name}' has shape {spec.shape}."
+            )
```

Comment:
Is this a breaking change?

```python
mlflow.pyfunc.spark_udf(
    "<model_uri>",  # your model's output a 3D tensor.
    # result_type unspecified
)
```

This code doesn't throw in mlflow <= 2.4.2, but it would in > 2.4.2.

Reply:
No. Previously it only support 1D array.

Reply:
and in previous mlflow version,
```
mlflow.pyfunc.spark_udf(
    "<model_uri>",  # your model's output a 3D tensor.
    # result_type unspecified
)
```
this is invalid code, because it set `result_type` to `double` which mismatches with model output schema.